lms/static/js/student_account/views/RegisterView.js

```diff
@@ -94,27 +94,35 @@
                 buildForm: function(data) {
                     var html = [],
                         i,
+                        field,
                         len = data.length,
                         requiredFields = [],
                         optionalFields = [];
 
                     this.fields = data;
 
+                    this.hasOptionalFields = false;
                     for (i = 0; i < len; i++) {
-                        if (data[i].errorMessages) {
+                        field = data[i];
+                        if (field.errorMessages) {
                             // eslint-disable-next-line no-param-reassign
-                            data[i].errorMessages = this.escapeStrings(data[i].errorMessages);
+                            field.errorMessages = this.escapeStrings(field.errorMessages);
                         }
 
-                        if (data[i].required) {
-                            requiredFields.push(data[i]);
+                        if (field.required) {
+                            requiredFields.push(field);
                         } else {
-                            optionalFields.push(data[i]);
+                            if (field.type !== 'hidden') {
+                                // For the purporse of displaying the optional field toggle,
+                                // the form should be considered to have optional fields
+                                // only if all of the optional fields are being rendering as
+                                // input elements that are visible on the page.
+                                this.hasOptionalFields = true;
```

Comment:
Most of the diff here is just assigning `data[i]` to the `field` variable and referencing `field` instead of indexing the array every time. This `if` statement is the meat of the change though.